chirho/robust/handlers/estimators.py

```diff
@@ -1,13 +1,95 @@
-from typing import Any, Callable, TypeVar
+import warnings
+from typing import Any, Callable
 
+import pyro
 import torch
-from typing_extensions import ParamSpec
 
-from chirho.robust.ops import Functional, Point, influence_fn
+from chirho.robust.ops import Functional, P, Point, S, T, influence_fn
 
-P = ParamSpec("P")
-S = TypeVar("S")
-T = TypeVar("T")
+
+class MonteCarloInfluenceEstimator(pyro.poutine.messenger.Messenger):
+    """
+    Effect handler for approximating efficient influence functions with nested monte carlo.
+    See the MC-EIF estimator in https://arxiv.org/pdf/2403.00158.pdf for details and
+    :func:`~chirho.robust.ops.influence_fn` for example usage.
+
+    .. note::
+
+        * ``functional`` must compose with ``torch.func.jvp``
+        * Since the efficient influence function is approximated using Monte Carlo, the result
+          of this function is stochastic, i.e., evaluating this function on the same ``points``
+          can result in different values. To reduce variance, increase ``num_samples_outer`` and
+          ``num_samples_inner`` in ``linearize_kwargs``.
+        * Currently, ``model`` cannot contain any ``pyro.param`` statements.
+          This issue will be addressed in a future release:
+          https://github.com/BasisResearch/chirho/issues/393.
+        * There are memory leaks when calling this function multiple times due to ``torch.func``.
+          See issue:
+          https://github.com/BasisResearch/chirho/issues/516.
+          To avoid this issue, use ``torch.no_grad()`` as shown in the example above.
+
+    """
+
+    def __init__(self, **linearize_kwargs):
+        self.linearize_kwargs = linearize_kwargs
+        super().__init__()
+
+    def _pyro_influence_fn(self, msg) -> None:
+        functional: Functional[P, S] = msg["args"][0]
+        points: Point[T] = msg["args"][1:]
```

Comment:
As the type checker error says, these type variables (`P`, `S` and `T`) are unbound. You can only use type variables in variable annotations if the type variables are introduced in an enclosing generic function type signature or `typing.Generic` type, so you need to either move the definition of `functional` and `points` inside `_influence_functional` (whose signature binds `P` and `S`), add `Generic[P, T, S]` as a parent class, or remove the variable annotations.

Reply:
Makes sense. Addressing your suggestion below simplified this substantially.